app/views/support/ABTest.scala

```diff
@@ -6,26 +6,51 @@ import java.net.URLEncoder
 import java.nio.charset.StandardCharsets
 
 import com.gu.i18n.CountryGroup
+import com.gu.i18n.CountryGroup._
 import play.api.libs.json._
 import play.api.mvc.{Cookie, Request}
 import play.twirl.api.Html
+import scala.util.Random
 
 import scalaz.NonEmptyList
+import views.html.fragments.giraffe.{contributeAmountButtons, contributeMessage}
 
 sealed trait TestTrait {
-
   type VariantFn
-  case class Variant(variantName: String, variantSlug: String, weight: Double, render: VariantFn) {
+  case class Variant(variantName: String, variantSlug: String, weight: Double, render: VariantFn, countryGroupFilter: Set[CountryGroup] = Set.empty) {
     def testName = name
     def testSlug = slug
+    def matches(countryGroup:CountryGroup):Boolean = countryGroupFilter.isEmpty || countryGroupFilter.contains(countryGroup)
   }
 
   def name: String
   def slug: String
   def variants: NonEmptyList[Variant]
-  val weight: Double = variants.map(_.weight).list.fold(0.0)(_ + _)
-  val cdf : Seq[Double]= variants.map(_.weight).list.foldLeft(Seq[Double]())( (l,p) => l :+ l.lastOption.getOrElse(0.0) + p/weight)
-  val weightedVariants = cdf.zip(variants.list)
+
+  def ValidateVariants = {
+    val countryWithNoVariants = variantsByCountry.collectFirst { case (country, variants) if variants.isEmpty => country.name }
+    val errorMessage = countryWithNoVariants.map(country => s"${getClass.getName}: No variant defined for $country.")
+
+    errorMessage.foreach(m => throw new IllegalStateException(m))
```

Comment:
> Is there a better way of dealing with errors rather than using exceptions? Could we enforce type safety so this fails at compile time rather than run time?


I think compile-time safety would be difficult, because the variants are defined with given countrygroups, and then you have to invert the map in order to find out which countries are not covered. Perhaps there is some kind of hardcore scalaz way of doing this, but I don't know what it is! However, it would be easy to create a unit test to ensure a bed setup fails at build time.